### ISSUE TITLE
Proposal: add `sourceContext` property to the messages generated by the `eslint` object

### ISSUE THREAD

**Issue by vitorbal** [eslint/eslint#7098]
**Context:**
We would like to port the JSCS default formatter over to ESLint. That formatter prints out the lines around the location of an error/warning. See #5860 for an example output.

**Problem**
Unfortunately, with ESLint, there is currently no way to get those lines of "context" from a formatter, because:

1. The formatter is invoked with only one argument - the array of messages to report on, and those do not contain any information about the lines of context around the error.
2. The formatter is intentionally made to be "dumb" and shouldn't/cannot use `SourceCode.getLines()` to retrieve the lines of code of the file being reported on.

**Proposal**
I would like to add a `sourceContext` property to each message object that gets returned by the `eslint.verify()` method.
With this new property, the message object would look roughly like the following:

```javascript
{
    ruleId: 'lines-around-comment',
    severity: 2,
    message: 'Expected line before comment.',
    line: 37,
    column: 9,
    nodeType: 'Line',
    source: '// some random comment',
    sourceContext: {
        before: [
            "line of code before line of the error",
            "another line of code"
        ],
        after: [
            "line of code after the line of the error",
            "another line of code"
        ]
    }
}
```

Each item of the `before`/`after` array is a line of code directly from the array returned by `SourceCode.getLines(code)`.

An alternative would be that each item is instead an object that contains a `source` and a `line` property, like this:

```javascript
{
    sourceContext: {
        before: [
            {
                source: "line of code before line of the error",
                line: 35
            },
            {
                source: "another line of code",
                line: 36
            }
        ],
        after: [
            // ...
        ]
    }
}
```
The advantage of this alternative approach is that it would prevent delegating the calculation of the line number to the formatter.

Adding this `sourceContext` property would solve the problem described, and would allow us to port the JSCS formatter with ease.
I have some proof-of-concept code for this proposal locally that is already able to print out the lines around an error, like this:

```bash
$ bin/eslint.js --rulesdir lib/internal-rules/ lib/formatters/stylish.js
35
36     results.forEach(function(result) {
37         // console.log(result);
38         const messages = result.messages;
39

lib/formatters/stylish.js
  37:9  error  Expected line before comment  lines-around-comment
```

The existing tests also already pass so I would only need to clean up the code a bit and make sure to test some edge cases better.

**Questions**
- Are there other any other ways to solve the problem described that would be preferable?
- I don't really like the property name `sourceContext` that much, but couldn't come up with anything better. Any suggestions?

**Comment by mikesherov** [comment]
The line number is already passed to Formatters. I think sourceContext should only have to be an array as per your first example. I also think sourceContext is a fine name considering git calls this context too.

One last approach, which might be overkill, is to pass the entire source text as a property on each message object. This would allow much more flexibility in Formatters to print out as much context as a formatter chooses

**Comment by IanVS** [comment]
>One last approach, which might be overkill, is to pass the entire source text as a property on each message object. This would allow much more flexibility in Formatters to print out as much context as a formatter chooses

I'd be curious about the performance impact of that.

**Comment by vitorbal** [comment]
So, I ran both versions in the eslint codebase, forcing ~4000 errors to show up, and it seems like passing the whole sourceCode with every message is quite memory intensive:

with no context: 190M
with a bit of context (max. 4 lines): 192M
with whole file as context: **236M**

IMO passing the whole source text might be overkill, for marginal value to formatters (I can't imagine a formatter wanting to print the entire source?).

What we could do is increase the number of context lines passed to give custom formatters some "wiggle room" in customization, say 5 lines before and 5 after. What do you guys think?

**Comment by IanVS** [comment]
>5 lines before and 5 after.

Personally I think that one line before and one after (or however much is needed to match JSCS reporter) should be enough to give the user context to the error in the code.  It could always be increased in the future if someone has a good pressing need for more.

**Comment by vitorbal** [comment]
Makes sense :+1:

**Comment by platinumazure** [comment]
Couldn't we leave it as a CLI option? By default return no extra context; if user specifies a number, return that many lines before/after; if user says "all" or similar, include full source.

Assumption is user can choose how much context they want, including none for formatters that don't even support it in the first place.

**Comment by btmills** [comment]
Is there a way for the formatter to take as much or as little as it wants? Not passing the whole source, but what about an API that gives it access to the source? (Similar to the `sourceCode` API.)

**Comment by vitorbal** [comment]
> Couldn't we leave it as a CLI option? By default return no extra context; if user specifies a number, return that many lines before/after; if user says "all" or similar, include full source.
> 
> Assumption is user can choose how much context they want, including none for formatters that don't even support it in the first place.

@platinumazure I think that could make a lot of sense if a few of our formatters could output some context depending on the presence or absence of a `sourceContext`. But wouldn't it be weird to have a CLI option that only works with certain formatters?

> Is there a way for the formatter to take as much or as little as it wants? Not passing the whole source, but what about an API that gives it access to the source? (Similar to the sourceCode API.)

@btmills I think the formatters were intentionally designed to be "dumb" and only export a function that takes in the array of problems and outputs them. I imagine giving them access to something similar to the `sourceCode` API would violate that design. How would you imagine that API being passed in? as an optional second argument?

**Comment by nzakas** [comment]
I'm not a fan of adding another property to each message, as it seems like the possibility of adding redundant information to multiple messages is quite high. Before I delve into this, I'd like to offer some background.

## Background

The way formatters currently work is that all of the linting is done first and the results are aggregated into a `report` object. Then, the `report` object is used to pass results into the formatter. This sequence is important because all `SourceCode` objects are destroyed by that point in time, and further, the execution context in which each individual file's linting was done is completely lost.

## Possible Solution

We could keep a map of filename to source code (plain text) as linting is being done. We'd only need entries in that map for files that had errors, so it could be kept relatively small. So basically, somewhere in `CLIEngine`, you'd end up with stuff like:

```js
var sourceCache = new Map();
sourceCache.set(filename, fs.readFileSync(filename, "utf8"));
```

Then, `sourceCache` could be passed into formatters as a second argument, so it's there if you want to grab stuff from it, but you're not required to use it.

**Comment by vitorbal** [comment]
@nzakas thank you for taking the time for the detailed response, really appreciate it!
I'm gonna give your proposed solution a try tonight and i'll report back :)

**Comment by nzakas** [comment]
@vitorbal sure thing, here's a hint on how i'd go about implementing it:

In `CLIEngine#getFormatter()`, I'd make that return a function that already has the cache bound to it. Something like this:

```js
CLIEngine.prototype.getFormatter = function(format) {
    // validationg stuff

   try {
        const formatter = require(formatterPath);
        return function(results) {
            return formatter(results, sourceCache);
        }
    } catch (ex) {

    }
};
```

Ugh, on second thought, it looks like we have a static `getFormatter()` method that wouldn't have access to a `sourceCache` object. :( So maybe my solution is a no-go.

**Comment by vitorbal** [comment]
@nzakas what if we add an additional property to the returned object of executeOnFiles(), like this: https://github.com/eslint/eslint/compare/jscs_formatter_v2?expand=1

**Comment by ilyavolodin** [comment]
@vitorbal One thing to keep in mind. We already had reports (couple of them a while back), that on a very large code base with a lot of errors, formatters run out of memory. If we add a lot of additional information to error code, those reports will most likely start showing up a lot more often.

**Comment by vitorbal** [comment]
@ilyavolodin that's a good point, thanks for mentioning it! I'll make sure we're not increasing the memory usage by much for the final solution we come up with.

**Comment by nzakas** [comment]
@vitorbal the downside of that approach is that the report object is no longer JSON-serializable.

We might just have to add a `source` property to the file result:

```
{
    results: [
        {
            filePath: "./myfile.js",
            source: "foo ;",
            output: "foo;",
            messages: [
                {
                    fatal: false,
                    severity: 2,
                    ruleId: "semi",
                    severity: 2,
                    line: 1,
                    column: 23,
                    message: "Expected a semicolon."
                }
            ],
            errorCount: 1,
            warningCount: 0
        }
    ],
    errorCount: 1,
    warningCount: 0
}
```

The total in-memory size would be the same as the `Map` approach, but would eliminate the need for an extra parameter to formatters.

We could further reduce the size of the results if we removed the `source` property from individual lint messages, which would become redundant at that point (though this would be a breaking change).

**Comment by vitorbal** [comment]
@nzakas that's a good point about keeping the report object JSON-serializable.

I tried the approach of adding a `source` property to the file result and it works fine!
I also did some rough before/after instrumentation in two different scenarios (26817 errors in 767 files, and 15584 errors in 279 files) and this doesn't really seem to cause any significant increase in memory usage.

Regarding the breaking change, I would suggest we do that part as a separate step. That way we can focus on porting the JSCS formatter first. What do you think?

**Comment by mikesherov** [comment]
:+1: from me on "no breaking change" as part of this PR.

**Comment by nzakas** [comment]
Fine by me. Let's just be sure to open another issue to track the breaking part.

**Comment by platinumazure** [comment]
@nzakas Do we have consensus to accept this issue (being a core change, not a rule change)? If not, should this be discussed at the next TSC meeting?

**Comment by nzakas** [comment]
## Design Summary

This change will be made in two phases:

1. Add a `source` property containing the entire file contents as a string into the linting results for each file that has linting errors. This can be used by formatters to output source code along with lint messages. This is a non-breaking change.
2. When we are ready for 4.0.0, we will remove the `source` property from individual linting error messages in order to avoid duplicate information in the results.

Just want to double-check that everyone is on the same page with this.

**Comment by nzakas** [comment]
Just one note that occurred to me: as we're planning on having 4.0.0 before the end of the year, it's probably best to keep the `source` property for each individual message until 5.0.0, to give people time to adjust their formatters.

**Comment by vitorbal** [comment]
Ok, sounds good. I'm gonna be making the follow-up issue to track those
changes in the coming days, was just a bit short on time lately!

**Comment by vitorbal** [comment]
Created a follow-up issue. You can find it here: https://github.com/eslint/eslint/issues/7358.
